File: server/test/random.test.js
/* */

import { use, expect } from "chai";
import chaiHttp from "chai-http";
const chai = use(chaiHttp);

import { aux } from "../middlewares.js";

const url = "http://localhost:3000";

describe("RANDOM API TESTS", () => {
  // Test para /v1/random/name
  describe("GET /v1/random/name", () => {
    it("should return a random name", (done) => {
      chai
        .request.execute(url)
        .get("/v1/random/name")
        .end((err, res) => {
          expect(res).to.have.status(200);
          expect(res.body).to.have.property("quest", "Random Name");
          expect(res.body.data).to.be.an("array").with.lengthOf(1);
          done();
        });
    });
    it("should return a 404 if only service is ok ", (done) => {
      chai
        .request.execute(url)
        .get("/v1/random")
        .end((err, res) => {
          expect(res).to.have.status(404);
          expect(res.body).to.have.property("msg", "Route Not Found");
          done();
        });
    });
    it("should return a bad request if service is ok but not the rest", (done) => {
      chai
        .request.execute(url)
        .get("/v1/random/inte")
        .end((err, res) => {
          expect(res).to.have.status(404);
          expect(res.body).to.have.property("msg", "Route Not Found");
          done();
        });
    });
  });

  // Test para /v1/random/integer
  describe("GET /v1/random/integer", () => {
    it("should return a random integer within the specified range", (done) => {
      chai
        .request.execute(url)
        .get("/v1/random/integer?min=1&max=10")
        .end((err, res) => {
          expect(res).to.have.status(200);
          expect(res.body).to.have.property("quest", "Random Integer between 1-10");
          expect(res.body.data).to.be.an("array").with.lengthOf(1);
          done();
        });
    });

    it("should return multiple random integers when times is provided", (done) => {
      chai
        .request.execute(url)
        .get("/v1/random/integer?min=1&max=10&times=5")
        .end((err, res) => {
          expect(res).to.have.status(200);
          expect(res.body).to.have.property("quest", "5 Random Integers between 1-10");
          expect(res.body.data).to.be.an("array").with.lengthOf(5);
          done();
        });
    });

    it("should return bad request if times <1 ", (done) => {
      chai
        .request.execute(url)
        .get("/v1/random/integer?min=1&max=10&times=-5")
        .end((err, res) => {
          expect(res).to.have.status(400);
          done();
        });
    });

    it("should return bad request if min is invalid", (done) => {
      chai
        .request.execute(url)
        .get("/v1/random/integer?min=-1&max=10")
        .end((err, res) => {
          expect(res).to.have.status(400);
          done();
        });
    });

    it("should return bad request if max is less than min", (done) => {
      chai
        .request.execute(url)
        .get("/v1/random/integer?min=10&max=1")
        .end((err, res) => {
          expect(res).to.have.status(400);
          done();
        });
    });
  });

  // Test para /v1/random/list
  describe("GET /v1/random/list", () => {
    it("should return a randomized list of specified length", (done) => {
      chai
        .request.execute(url)
        .get("/v1/random/list?len=5")
        .end((err, res) => {
          expect(res).to.have.status(200);
          expect(res.body).to.have.property("quest", "Randomized order list with 5 elements");
          expect(res.body.data).to.be.an("array").with.lengthOf(5);
          done();
        });
    });

    it("should return bad request if list length is too small", (done) => {
      chai
        .request.execute(url)
        .get("/v1/random/list?len=1")
        .end((err, res) => {
          expect(res).to.have.status(400);
          done();
        });
    });

    it("should return bad request if list length is too large", (done) => {
      chai
        .request.execute(url)
        .get("/v1/random/list?len=10001")
        .end((err, res) => {
          expect(res).to.have.status(400);
          done();
        });
    });
  });
});

File: random/random_test.go
package random

import (
	"encoding/json"
	"fmt"
	"net/http"
	"testing"
)

// Estructura para manejar la respuesta JSON de la API
type RandomTestOutput struct {
	Quest string `json:"quest,omitempty"`
	Data  []int  `json:"data,omitempty"`
	Msg   string `json:"msg,omitempty"`
	Error string `json:"Error,omitempty"`
}

func TestRandomApi(t *testing.T) {
	const (
		serverURL   = "http://localhost:3000"
		validFormat = "Bad request, valid format is 'api.codetabs.com/v1/{service}?{param}=value'. Please read our docs at https://codetabs.com"
	)

	var randomTests = []struct {
		it         string
		endpoint   string
		statusCode int
		quest      string
		dataLen    int
		msg        string
		errorText  string
	}{
		// Tests /v1/random/name
		{"get random name", "/v1/random/name", 200,
			"Random Name", 0, "", ""},
		{"404 route not found", "/v1/random", 400,
			"", 0, "", validFormat},
		{"bad request for invalid random endpoint", "/v1/random/inte", 400,
			"", 0, "", validFormat},

		// Tests /v1/random/integer
		{"get random integer", "/v1/random/integer?min=1&max=10", 200,
			"Random Integer between 1-10", 1, "", ""},
		{"get multiple integers", "/v1/random/integer?min=1&max=10&times=5", 200,
			"5 Random Integers between 1-10", 5, "", ""},
		{"bad request times <1", "/v1/random/integer?min=1&max=10&times=-5", 400,
			"", 0, "", validFormat},
		{"bad request invalid min", "/v1/random/integer?min=-1&max=10", 400,
			"", 0, "", validFormat},
		{"bad request max less than min", "/v1/random/integer?min=10&max=1", 400, "", 0, "", validFormat},

		// Tests /v1/random/list
		{"get randomized list", "/v1/random/list?len=5", 200, "Randomized order list with 5 elements", 5, "", ""},
		{"bad request for list length too small", "/v1/random/list?len=1", 400, "", 0, "", validFormat},
		{"bad request for list length too large", "/v1/random/list?len=10001", 400, "", 0, "", validFormat},
	}

	for _, test := range randomTests {
		var to RandomTestOutput

		req, err := http.NewRequest("GET", fmt.Sprintf("%s%s", serverURL, test.endpoint), nil)
		if err != nil {
			t.Errorf("Error preparing request for %s: %v", test.it, err)
			continue
		}

		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			t.Fatalf("Error making request for %s: %v", test.it, err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != test.statusCode {
			t.Errorf("%s got status %v, expected %v", test.it, resp.StatusCode, test.statusCode)
		}

		err = json.NewDecoder(resp.Body).Decode(&to)
		if err != nil {
			t.Fatalf("Error decoding response for %s: %v", test.it, err)
		}

		if resp.StatusCode == 200 && to.Quest != test.quest {
			t.Errorf("%s got quest '%v', expected '%v'", test.it, to.Quest, test.quest)
		}

		if resp.StatusCode == 200 && len(to.Data) != test.dataLen {
			t.Errorf("%s got data length %v, expected %v", test.it, len(to.Data), test.dataLen)
		}

		if resp.StatusCode == 400 && to.Error != test.errorText {
			t.Errorf("%s got error '%v', expected '%v'", test.it, to.Error, test.errorText)
		}

		if test.msg != "" && to.Msg != test.msg {
			t.Errorf("%s got msg '%v', expected '%v'", test.it, to.Msg, test.msg)
		}

		fmt.Printf("Test OK... %s\n", test.it)
	}
}
